### PR TITLE
pinned comments

### PR DESCRIPTION
**Pull request by SirSaltyy**
Didn't test it yet since my local environment still scuffed.

So the pinned comment UI might need iterating on.

### CODE REVIEW

**Comment by jahooma** [comment]
Oh, I guess I should review the code

File: web/components/feed/feed-comments.tsx [hunk 1]
```diff
@@ -334,6 +334,9 @@ export const FeedComment = memo(function FeedComment(props: {
           />
 
           <HideableContent comment={comment} />
+          <PinnableContent comment={comment}>
+            <Col className="group"></Col>
+          </PinnableContent>
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:339]: These lines were just for testing right?

File: web/components/contract/contract-tabs.tsx [hunk 2]
```diff
@@ -364,54 +365,66 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
+            onSubmitReply={loadNewer}
+            bets={bets?.filter(
+              (b) =>
+                b.replyToCommentId &&
+                [parent]
+                  .concat(commentsByParent[parent.id] ?? [])
+                  .map((c) => c.id)
+                  .includes(b.replyToCommentId)
+            )}
+          />
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:396]: Duplicating this code for the feed thread is not ideal.

You can simply insert the pinned comments first in the parent comments array. Ask chat gpt

File: web/components/comments/comments-list.tsx [hunk 3]
```diff
@@ -75,13 +74,21 @@ export function UserCommentsList(props: { user: User }) {
...
+            <div key={i}>
+              {pinnedComments.map((comment) => (
+                <ProfileComment key={comment.id} comment={comment} />
+              ))}
+
+              {/* Then render regular comments */}
+              {regularComments.map((comment) => (
+                <ProfileComment key={comment.id} comment={comment} />
+              ))}
+            </div>
```
Comment (jahooma) [inline web/components/comments/comments-list.tsx:91]: Isn't this a totally different meaning of pinned, if it is pinned not only on the market, but on your comments list?

Seems like you should just do without this bit.

File: backend/api/src/pin-comment.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import * as admin from 'firebase-admin'
+import { Contract } from 'common/contract'
+import { isAdminId, isModId } from 'common/envs/constants'
+import { getUser, revalidateContractStaticProps } from 'shared/utils'
+import { APIError, type APIHandler } from './helpers/endpoint'
+
+export const pinComment: APIHandler<'pin-comment'> = async (
+  { commentPath },
```
Comment (jahooma) [inline backend/api/src/pin-comment.ts:8]: Hmm, I think this is a lot more confusing than just passing the `contractId` and `commentId` and reconstructing the path here.

File: web/components/contract/contract-tabs.tsx [hunk 5]
```diff
@@ -364,54 +365,66 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
-            firebaseLogin,
-            'sign up to comment button click'
-          )}
-          className={clsx('mt-4', comments.length > 0 && 'ml-12')}
-          size="lg"
-          color="gradient"
-        >
-          Sign up to comment <ArrowRightIcon className="ml-2 h-4 w-4" />
-        </Button>
-      )}
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:414]: Uh, I don't think you want to delete this code?

**Review by jahooma — CHANGES_REQUESTED** [review]
Nice start on this feature!

Just a few comments for you.

File: web/components/comments/comments-list.tsx (hunk 3 above)
Comment (SirSaltyy) [reply web/components/comments/comments-list.tsx:91]: whoops, undid all changes to this file in my current code

File: backend/api/src/pin-comment.ts (hunk 4 above)
Comment (SirSaltyy) [reply backend/api/src/pin-comment.ts:8]: I was just doing the same method we currently use for hide-comment.ts

Still don't fully understand backend stuff so just trying to stick to our current code. Very happy for you to explain further though.

File: web/components/contract/contract-tabs.tsx (hunk 5 above)
Comment (SirSaltyy) [reply web/components/contract/contract-tabs.tsx:414]: lol good catch, have re-added it to my current code.

File: web/components/contract/contract-tabs.tsx (hunk 2 above)
Comment (SirSaltyy) [reply web/components/contract/contract-tabs.tsx:396]: Before seeing this I had already made some changes. I'm not inserting them into the parent comments array, but after reviewing my new code let me know if I should still make this change.

File: web/components/feed/feed-comments.tsx (hunk 1 above)
Comment (SirSaltyy) [reply web/components/feed/feed-comments.tsx:339]: uhh no haha. I don't think so?

This is the code that renders the 📌 and in the next version the "view context" link.

File: web/components/contract/contract-tabs.tsx [hunk 6]
```diff
@@ -319,10 +340,22 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
   }, [idToHighlight, comments.length])
 
   const loadMore = () => setParentCommentsToRender((prev) => prev + LOAD_MORE)
+  const pinnedComments = comments.filter((comment) => comment.pinned)
   const onVisibilityUpdated = useEvent((visible: boolean) => {
     if (visible) loadMore()
   })
 
+  const [idInUrl, setIdInUrl] = useState<string | undefined>(undefined)
+  const hash = useHash()
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:349]: You can just use `useHashInUrl` to get the hash.

File: web/components/contract/contract-tabs.tsx [hunk 7]
```diff
@@ -40,9 +40,30 @@ import { firebaseLogin } from 'web/lib/firebase/users'
 import { ArrowRightIcon } from '@heroicons/react/outline'
 import clsx from 'clsx'
 import { useRealtimeCommentsOnContract } from 'web/hooks/use-comments-supabase'
+import { ParentFeedComment, FeedComment } from '../feed/feed-comments'
+import { scrollIntoViewCentered } from 'web/lib/util/scroll'
 
 export const EMPTY_USER = '_'
 
+function useHash() {
```
[the remark this answers is not in the document]
Comment (jahooma) [reply web/components/contract/contract-tabs.tsx:48]: Actually you can delete this code and use `useHashInUrl`

File: web/components/contract/contract-tabs.tsx [hunk 8]
```diff
@@ -364,34 +398,50 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
+      {comments
+        .filter((comment) => !comment.replyToCommentId)
+        .map((parent) => (
+          <FeedCommentThread
+            key={parent.id}
+            contract={contract}
+            parentComment={parent}
+            threadComments={commentsByParent[parent.id] ?? []}
+            trackingLocation={'contract page'}
+            idInUrl={idInUrl}
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:425]: I think if you pass `idToHighlight` here, you can delete the code for `idInUrl`

File: web/components/contract/contract-tabs.tsx [hunk 9]
```diff
@@ -364,34 +398,50 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
+            seeReplies={false}
+            numReplies={0}
+            pinnedHighlighted={true}
+            setIdInUrl={setIdInUrl}
+          />
+        </div>
       ))}
 
+      {comments
+        .filter((comment) => !comment.replyToCommentId)
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:417]: Use `parentComments` instead of this

Comment (jahooma) [reply web/components/contract/contract-tabs.tsx:417]: Use `parentComments.slice(0, parentCommentsToRender)`

File: web/components/contract/contract-tabs.tsx [hunk 10]
```diff
@@ -416,6 +466,76 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
+  } = props
+
+  const handleContextClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
+    e.preventDefault()
+    const commentId = comment.id
+    setIdInUrl(commentId)
+
+    const currentUrl = window.location.href
+    const newUrl = currentUrl.split('#')[0] + `#${commentId}`
+    window.history.pushState({}, '', newUrl)
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:497]: Change this to `Router.push(newUrl)`

Comment (jahooma) [reply web/components/contract/contract-tabs.tsx:497]: Import Router:

`import Router from 'next/router`

File: web/components/feed/feed-comments.tsx [hunk 11]
```diff
@@ -320,21 +328,25 @@ export const FeedComment = memo(function FeedComment(props: {
...
+              contract={contract}
+              inTimeline={inTimeline}
+              isParent={isParent}
+            />
+            {pin}
+          </Row>
 
           <HideableContent comment={comment} />
-          <Row>
+          <Row className="flex flex-wrap items-start">
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:348]: `flex` is already included on `Row` components

File: web/components/feed/feed-comments.tsx [hunk 12]
```diff
@@ -396,18 +409,21 @@ export const FeedComment = memo(function FeedComment(props: {
...
   onReplyClick?: (comment: ContractComment) => void
-  onSeeReplyClick: () => void
+  onSeeReplyClick?: () => void
   trackingLocation: string
   inTimeline?: boolean
   childrenBountyTotal?: number
   bets?: Bet[]
+  viewContext?: ReactNode
+  pinnedHighlighted?: boolean
+  pin?: ReactNode
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:426]: `isPinned?: boolean`

Then configure all within ParentFeedComment.

File: web/components/feed/feed-comments.tsx [hunk 13]
```diff
@@ -211,6 +211,9 @@ export const FeedComment = memo(function FeedComment(props: {
   isParent?: boolean
   bets?: Bet[]
   lastInReplyChain?: boolean
+  viewContext?: ReactNode
+  pinnedHighlighted?: boolean
+  pin?: ReactNode
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:216]: `isPinned?: boolean`

File: web/components/feed/feed-comments.tsx [hunk 14]
```diff
@@ -320,21 +328,25 @@ export const FeedComment = memo(function FeedComment(props: {
...
-          />
+          <Row className="flex items-center justify-between">
+            <FeedCommentHeader
+              comment={comment}
+              updateComment={updateComment}
+              contract={contract}
+              inTimeline={inTimeline}
+              isParent={isParent}
+            />
+            {pin}
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:344]: Replace with:
```
{isPinned & (
<div>...</div>
)}
```

File: web/components/feed/feed-comments.tsx [hunk 15]
```diff
@@ -320,21 +328,25 @@ export const FeedComment = memo(function FeedComment(props: {
...
+          </Row>
 
           <HideableContent comment={comment} />
-          <Row>
+          <Row className="flex flex-wrap items-start">
+            {viewContext}
```
Comment (jahooma) [inline web/components/feed/feed-comments.tsx:349]: Similar

File: web/components/contract/contract-tabs.tsx [hunk 16]
```diff
@@ -416,6 +466,76 @@ export const CommentsTabContent = memo(function CommentsTabContent(props: {
...
+    seeReplies,
+    numReplies,
+    pinnedHighlighted,
+    setIdInUrl,
+  } = props
+
+  const handleContextClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
+    e.preventDefault()
+    const commentId = comment.id
+    setIdInUrl(commentId)
```
Comment (jahooma) [inline web/components/contract/contract-tabs.tsx:493]: Could try deleting this, seems like it should work if we are updating the hash (and listening for hash changes above)

**Review by jahooma — COMMENTED** [review]
Nice! Almost there. Just a few quibbles haha

File: web/components/contract/contract-tabs.tsx (hunk 16 above)
Comment (SirSaltyy) [reply web/components/contract/contract-tabs.tsx:493]: WRONG

**Pull request merged**